ambry-clustermap/src/main/java/com.github.ambry.clustermap/DynamicClusterChangeHandler.java

```diff
@@ -360,10 +363,20 @@ private void addOrUpdateInstanceInfos(List<InstanceConfig> instanceConfigs) thro
             // Ensure only one AmbryPartition instance exists for specific partition.
             mappedPartition = clusterChangeHandlerCallback.addPartitionIfAbsent(mappedPartition, replicaCapacity);
             ensurePartitionAbsenceOnNodeAndValidateCapacity(mappedPartition, dataNode, replicaCapacity);
-            // create new replica belonging to this partition
-            AmbryReplica replica =
-                new AmbryServerReplica(clusterMapConfig, mappedPartition, disk, stoppedReplicas.contains(partitionName),
-                    replicaCapacity, sealedReplicas.contains(partitionName));
+            // create new replica belonging to this partition or find the existing replica from bootstrapReplicas map.
+            AmbryReplica replica;
+            if (selfInstanceName.equals(instanceName)) {
+              // if this is a newly added replica on current instance, it should be present in bootstrapReplicas map.
+              replica = (AmbryReplica) bootstrapReplicas.remove(mappedPartition.toPathString());
+              if (replica == null) {
+                logger.warn("Replica {} is not present in bootstrap replica set, skip adding it to clustermap",
```

Comment:
Would it be worth having a metric for this event? Would this mean that the replica never gets added until the machine is restarted or will it get added eventually?

Reply:
Actually, `replica == null` should not happen in practice. If everything goes well, then the replica should be present in bootstrap replica map. If failure occurs during bootstrap, there are two cases to discuss:
(1)  new replica is added but node crashes before updating Helix InstanceConfig
       There is no InstanceConfig change. When machine is restarted, node will re-do replica addition and newly added replica should be in the map.
(2) InstanceConfig is updated and then node crashes immediately
       Since new replica is already in InstanceConfig, once machine is restarted the new replica will be added to in-mem clustermap when handling initial InstanceConfig change (a different code path within `createNewInstance()` method)

There is only one rare case (human error) where we mistakenly update InstanceConfig of existing node by adding new replica rather than update IdealState, which may trigger this `replica == null` statement.